prog/target.go

```diff
@@ -80,6 +80,9 @@ type Target struct {
 	// The default ChoiceTable is used only by tests and utilities, so we initialize it lazily.
 	defaultOnce        sync.Once
 	defaultChoiceTable *ChoiceTable
+
+	hintAttemptsMu sync.Mutex
+	hintAttempts   map[uint64]int
```

Comment:
Would it be more clean/less surprising if we explicitly passed it to `prog.Hints` instead of hiding it in `Target`?

Or, probably, it is also okay to restrict the number of hints per PC only per a single hint mutation? I guess, it should already do most of the reduction?

Reply:
> Would it be more clean/less surprising if we explicitly passed it to prog.Hints instead of hiding it in Target?

Do you mean to pass it to prog.MutateWithHints?
Then we will also need to pass the mutex, which is a bit strange.

Reply:
> Or, probably, it is also okay to restrict the number of hints per PC only per a single hint mutation? I guess, it should already do most of the reduction?

I am not so sure.
Say, if we any openat syscall (for all kinds of devices), it passes through common open-related code and does a bunch of comparisons of flags/mode and probably more. All of these will be duplicates lots of times.
The same stands for most syscalls, e.g. any syscall that accepts and fd calls fd_get-related code which has a bunch of comparisons (uninteresting, e.g .current file table size).
I wanted to specifically limit across all hint mutations, so that we slowly phase out these common ones and do only new PCs.

Reply:
> Then we will also need to pass the mutex, which is a bit strange.

We could use `sync.Map` or hide the mutex in some custom structure.

Reply:
Regarding using the map for one hint mutation only -- I remember, the biggest hint generators were fs images (+a few networking calls?) The others did not have nearly as impressive numbers of hint substitutions. So I guess that even limiting those big ones can already offer most of the impact of this commit.

Reply:
There are dozen syz_mount_image syscalls, but 4 thousands other syscalls. Even instances that don't have syz_mount_image enabled don't cope with hints jobs now.

Reply:
I've removed hints state from prog.Target and instead added separate type HintsLimiter and added it to pkg/fuzzer.
PTAL